src/main/java/rx/internal/operators/OperatorMerge.java

```diff
@@ -71,715 +72,878 @@
+            }
+            if (success) {
+                boolean skipFinal = false;
+                try {
+                    try {
+                        child.onNext(value);
+                    } catch (Throwable t) {
+                        skipFinal = true;
+                        Exceptions.throwIfFatal(t);
+                        try {
+                            child.onError(t);
+                        } finally {
+                            unsubscribe();
+                        }
+                        return;
                     }
-                } else if (r > 0) {
-                    // drain what was requested
-                    long toEmit = r;
-                    for (int i = 0; i < toEmit; i++) {
-                        Object o = svq.poll();
-                        if (o == null) {
-                            break;
-                        } else {
-                            on.accept(actual, o);
-                            emittedWhileDraining++;
+                    producer.produced(1);
+                    request(1);
+                    // check if some state changed while emitting
+                    synchronized (this) {
+                        skipFinal = true;
+                        if (!missed) {
+                            emitting = false;
+                            return;
                         }
+                        missed = false;
                     }
-                    // decrement the number we emitted from outstanding requests
-                    MergeProducer.REQUESTED.getAndAdd(mergeProducer, -emittedWhileDraining);
-                }
-                return emittedWhileDraining;
-            }
-            return 0;
-        }
-
-        final Func1<InnerSubscriber<T>, Boolean> DRAIN_ACTION = new Func1<InnerSubscriber<T>, Boolean>() {
-
-            @Override
-            public Boolean call(InnerSubscriber<T> s) {
-                if (s.q != null) {
-                    long r = mergeProducer.requested;
-                    int emitted = s.drainQueue();
-                    if (emitted > 0) {
-                        s.requestMore(emitted);
+                } finally {
+                    if (!skipFinal) {
+                        synchronized (this) {
+                            emitting = false;
+                        }
                     }
-                    if (emitted == r) {
-                        // we emitted as many as were requested so stop the forEach loop
-                        return Boolean.FALSE;
+                }
+                // in case there was some activity, continue with the regular emission loop
+                emitLoop();
+            } else {
+                // either someone else is emitting or there was no available request
+                // queue up the scalar
+                RxRingBuffer svq = scalarQueue;
+                if (svq == null) {
+                    svq = RxRingBuffer.getSpscInstance();
```

Comment:
> Even if say its thread safe, you have to flip back and forth to verify it.

Yes, of course. But docs then show that this is considered. As it stands it is right next to code that must handle concurrency so this code stands out as being wrong.